File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeProviderRouter.kt
package io.github.rajibsarwar.edgeframeworks

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class EdgeProviderRouter(
    providers: List<EdgeModelProvider> = emptyList()
) {
    private val mutex = Mutex()
    private val providers = providers.toMutableList()

    suspend fun register(provider: EdgeModelProvider) {
        mutex.withLock {
            providers.removeAll { it.id == provider.id }
            providers.add(provider)
        }
    }

    suspend fun unregister(id: String) {
        mutex.withLock {
            providers.removeAll { it.id == id }
        }
    }

    suspend fun provider(
        requiredCapabilities: Set<EdgeCapability>
    ): EdgeModelProvider? {
        val snapshot = mutex.withLock { providers.toList() }

        for (provider in snapshot) {
            val capabilities = provider.capabilities()
            if (capabilities.containsAll(requiredCapabilities)) {
                return provider
            }
        }

        return null
    }
}

File: ios/Sources/EdgeFrameworks/Runtime/EdgeProviderRouter.swift
import Foundation

public actor EdgeProviderRouter {
    private var providers: [any EdgeModelProvider]

    public init(providers: [any EdgeModelProvider] = []) {
        self.providers = providers
    }

    public func register(_ provider: any EdgeModelProvider) {
        providers.removeAll { $0.id == provider.id }
        providers.append(provider)
    }

    public func unregister(id: String) {
        providers.removeAll { $0.id == id }
    }

    public func provider(
        supporting requiredCapabilities: Set<EdgeCapability>
    ) async -> (any EdgeModelProvider)? {
        for provider in providers {
            let capabilities = await provider.capabilities()
            if requiredCapabilities.isSubset(of: capabilities) {
                return provider
            }
        }

        return nil
    }
}
